### ISSUE TITLE
Rule Change: ignore URLs in comments only if they're the only content in the line

### ISSUE THREAD

**Issue by piranna** [eslint/eslint#15928]
### What rule do you want to change?

max-len

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
// http://example.com/a-very-long-url-as-single-item-in-a-comment-should-pass

// If comment has more text, the http://example.com/very-long-url
// would fail and should need to be moved to a new line
```

### What does the rule currently do for this code?

if `ignoreUrls` is enabled and line has an URL, full line is ignored, no matter if it's on code or a comment or has some extra text.

### What will the rule do after it's changed?

Only ignore lines that has a comment with an URL as its only content. If it's a string in code, or a trailling comment, or comment line has more text, it should fail, forcing you to split it to adjust to the `max-len` limit.

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

It's a more strict version of `ignoreUrls` option, since URLs in comments are mostly used as references so people can navigate to them, while long URLs in code can be splitted in multiple fragments, increasing readibility and reusability, and more important stick to the `max-len` provided limit.

**Comment by mdjermanovic** [comment]
Hi @piranna, thanks for the issue! 

Per our rules policies, [stylistic rules are frozen](https://github.com/eslint/eslint#stylistic-rule-updates), so we are not adding new options to the `max-len` rule. You can always copy the existing rule into a custom rule and make it behave the way you want.

> Only ignore lines that has a comment with an URL as its only content. If it's a string in code, or a trailling comment, or comment line has more text, it should fail, forcing you to split it to adjust to the `max-len` limit.

For this particular case, I think you could also use the existing option [`ignorePattern`](https://eslint.org/docs/rules/max-len#ignorepattern): set the pattern to match those lines, and turn off `ignoreUrls`. For example:

```js
/* eslint max-len: ["error", {
    code: 60,
    ignoreUrls: false,
    ignorePattern: "(?:\\s*)//\\s*\\S*[^:/?#]://[^?#]\\S*$"
}]*/


// http://example.com/a-very-long-url-as-single-item-in-a-comment-should-pass

// If comment has more text, the http://example.com/very-long-url
// would fail and should need to be moved to a new line
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG1heC1sZW46IFtcImVycm9yXCIsIHtcbiAgICBjb2RlOiA2MCxcbiAgICBpZ25vcmVVcmxzOiBmYWxzZSxcbiAgICBpZ25vcmVQYXR0ZXJuOiBcIig/OlxcXFxzKikvL1xcXFxzKlxcXFxTKlteOi8/I106Ly9bXj8jXVxcXFxTKiRcIlxufV0qL1xuXG5cbi8vIGh0dHA6Ly9leGFtcGxlLmNvbS9hLXZlcnktbG9uZy11cmwtYXMtc2luZ2xlLWl0ZW0taW4tYS1jb21tZW50LXNob3VsZC1wYXNzXG5cbi8vIElmIGNvbW1lbnQgaGFzIG1vcmUgdGV4dCwgdGhlIGh0dHA6Ly9leGFtcGxlLmNvbS92ZXJ5LWxvbmctdXJsXG4vLyB3b3VsZCBmYWlsIGFuZCBzaG91bGQgbmVlZCB0byBiZSBtb3ZlZCB0byBhIG5ldyBsaW5lIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLWxvc3Mtb2YtcHJlY2lzaW9uIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1ub25vY3RhbC1kZWNpbWFsLWVzY2FwZSI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNldHRlci1yZXR1cm4iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW5zYWZlLW9wdGlvbmFsLWNoYWluaW5nIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1iYWNrcmVmZXJlbmNlIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnt9fX0=)

(`(?:)` is only to avoid `*/`, which would end the configuration comment; `[^:/?#]://[^?#]` part is copied from [`URL_REGEXP`](https://github.com/eslint/eslint/blob/10249ad1a961463b6b347be71c074951ab8e2652/lib/rules/max-len.js#L98) in the `max-len` rule).

**Comment by piranna** [comment]
> For this particular case, I think you could also use the existing option [`ignorePattern`](https://eslint.org/docs/rules/max-len#ignorepattern)

In fact, I was thinking on using `ignorePattern` as an alternative :-D Just only I was worried about how to detect the URLs inside the multi-line comments, that's why I didn't tried it first. Do you have any clue? As you've showed, for the trailling comments syntax, regular expression to detect them is trivial :-)


> `(?:)` is only to avoid `*/`, which would end the configuration comment

So this is only just due to being set the eslint rule as a comment for example purposses, and at `.eslintrc` file it would not be needed, isn't it? Good catch :-)

**Comment by mdjermanovic** [comment]
> In fact, I was thinking on using `ignorePattern` as an alternative :-D Just only I was worried about how to detect the URLs inside the multi-line comments, that's why I didn't tried it first. Do you have any clue? As you've showed, for the trailling comments syntax, regular expression to detect them is trivial :-)

multi-line comments where the URL is on a separate line, like this one?

```js
/*
    http://example.com/a-very-long-url-as-single-item-in-a-comment-should-pass
*/
```

> So this is only just due to being set the eslint rule as a comment for example purposses, and at `.eslintrc` file it would not be needed, isn't it? Good catch :-)

Exactly, in a configuration file it can be `"\\s*//\\s*\\S*[^:/?#]://[^?#]\\S*$"`, there's no need to wrap the first part in `(?:)`.

**Comment by piranna** [comment]
> multi-line comments where the URL is on a separate line, like this one?

Yes, like that one, maybe with a `*` at begin of the line for aesthetics, nothing else.


> Exactly, in a configuration file it can be `"\\s*//\\s*\\S*[^:/?#]://[^?#]\\S*$"`, there's no need to wrap the first part in `(?:)`.

Great, thank you :-) Just only, with the final `$` at `[^?#]\\S*$`, that would prevent to have any anchor or query in the URL, isn't it? Maybe we should remove that part and just only left the schema and the `://` fragment?

**Comment by mdjermanovic** [comment]
> Great, thank you :-) Just only, with the final `$` at `[^?#]\\S*$`, that would prevent to have any anchor or query in the URL, isn't it? Maybe we should remove that part and just only left the schema and the `://` fragment?

It's to disallow additional text after the URL, for example:

```js
    // http://example.com/a-very-long-url-as-single-item-in-a-comment-should-pass some text that should be on the next line
```

`\\S*$` ensures that all characters after `://` belong to the url, as it would not match if there is a space.

Here's a [demo example](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG1heC1sZW46IFtcImVycm9yXCIsIHtcbiAgICBjb2RlOiA2MCxcbiAgICBpZ25vcmVVcmxzOiBmYWxzZSxcbiAgICBpZ25vcmVQYXR0ZXJuOiBcIig/OlxcXFxzKikvL1xcXFxzKlxcXFxTKlteOi8/I106Ly9bXj8jXVxcXFxTKiRcIlxufV0qL1xuXG5cbi8vIGh0dHA6Ly9leGFtcGxlLmNvbS9hLXZlcnktbG9uZy11cmwtYXMtc2luZ2xlLWl0ZW0taW4tYS1jb21tZW50LXNob3VsZC1wYXNzXG5cbi8vIElmIGNvbW1lbnQgaGFzIG1vcmUgdGV4dCwgdGhlIGh0dHA6Ly9leGFtcGxlLmNvbS92ZXJ5LWxvbmctdXJsXG4vLyB3b3VsZCBmYWlsIGFuZCBzaG91bGQgbmVlZCB0byBiZSBtb3ZlZCB0byBhIG5ldyBsaW5lXG5cbi8vIGh0dHA6Ly9leGFtcGxlLmNvbS92ZXJ5LWxvbmctdXJsIHdpdGggc29tZSB0ZXh0IGFmdGVyIHNob3VsZCBub3QgcGFzc1xuXG4vLyBodHRwOi8vZXhhbXBsZS5jb20vdmVyeS1sb25nLXVybD93aXRoPXNvbWUtcXVlcnkjYW5kLWFuY2hvci1zaG91bGQtcGFzcyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWVsc2UtaWYiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWltcG9ydC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1sb3NzLW9mLXByZWNpc2lvbiI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tbm9ub2N0YWwtZGVjaW1hbC1lc2NhcGUiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zZXR0ZXItcmV0dXJuIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVuc2FmZS1vcHRpb25hbC1jaGFpbmluZyI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtYmFja3JlZmVyZW5jZSI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7fX19).

`[^?#]` just disallows `?` or `#` as the first character after `://`, as it would be invalid url (and therefore assumed as not an url).

**Comment by mdjermanovic** [comment]
> > multi-line comments where the URL is on a separate line, like this one?
> 
> Yes, like that one, maybe with a `*` at begin of the line for aesthetics, nothing else.

Here's a version that allows urls after `//` or `*`:

```js
/* eslint max-len: ["error", {
    code: 60,
    ignoreUrls: false,
    ignorePattern: "^\\s*(?://|\\*)\\s*\\S*[^:/?#]://[^?#]\\S*$"
}]*/

// http://example.com/a-very-long-url-as-single-item-in-a-comment-should-pass

/*
 * http://example.com/a-very-long-url-as-single-item-in-a-comment-should-pass
 */

/*
 * http://example.com/a-very-long-url-as-single-item-in-a-comment-should-pass
 * this text is ok
 */

// with text before should not pass http://example.com/a-very-long-url

// http://example.com/a-very-long-url with text after should not pass 

/*
 * with text before should not pass http://example.com/a-very-long-url
 */

/*
 * http://example.com/a-very-long-url with text after should not pass 
 */

```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG1heC1sZW46IFtcImVycm9yXCIsIHtcbiAgICBjb2RlOiA2MCxcbiAgICBpZ25vcmVVcmxzOiBmYWxzZSxcbiAgICBpZ25vcmVQYXR0ZXJuOiBcIl5cXFxccyooPzovL3xcXFxcKilcXFxccypcXFxcUypbXjovPyNdOi8vW14/I11cXFxcUyokXCJcbn1dKi9cblxuLy8gaHR0cDovL2V4YW1wbGUuY29tL2EtdmVyeS1sb25nLXVybC1hcy1zaW5nbGUtaXRlbS1pbi1hLWNvbW1lbnQtc2hvdWxkLXBhc3NcblxuLypcbiAqIGh0dHA6Ly9leGFtcGxlLmNvbS9hLXZlcnktbG9uZy11cmwtYXMtc2luZ2xlLWl0ZW0taW4tYS1jb21tZW50LXNob3VsZC1wYXNzXG4gKi9cblxuLypcbiAqIGh0dHA6Ly9leGFtcGxlLmNvbS9hLXZlcnktbG9uZy11cmwtYXMtc2luZ2xlLWl0ZW0taW4tYS1jb21tZW50LXNob3VsZC1wYXNzXG4gKiB0aGlzIHRleHQgaXMgb2tcbiAqL1xuXG4vLyB3aXRoIHRleHQgYmVmb3JlIHNob3VsZCBub3QgcGFzcyBodHRwOi8vZXhhbXBsZS5jb20vYS12ZXJ5LWxvbmctdXJsXG5cbi8vIGh0dHA6Ly9leGFtcGxlLmNvbS9hLXZlcnktbG9uZy11cmwgd2l0aCB0ZXh0IGFmdGVyIHNob3VsZCBub3QgcGFzcyBcblxuLypcbiAqIHdpdGggdGV4dCBiZWZvcmUgc2hvdWxkIG5vdCBwYXNzIGh0dHA6Ly9leGFtcGxlLmNvbS9hLXZlcnktbG9uZy11cmxcbiAqL1xuXG4vKlxuICogaHR0cDovL2V4YW1wbGUuY29tL2EtdmVyeS1sb25nLXVybCB3aXRoIHRleHQgYWZ0ZXIgc2hvdWxkIG5vdCBwYXNzIFxuICovXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

**Comment by piranna** [comment]
> `\\S*$` ensures that all characters after `://` belong to the url, as it would not match if there is a space.

You are right, we want the line end there and don't have more values... I was not sure if `max-len` was checking just for the URL or the full line, now it makes more sense, thanks.


> Here's a version that allows urls after `//` or `*`:

Good point :-D I was worried about how to detect the `*`, but seems it's only possible when inside a comment or a multiline string template, anything else `eslint` gives an error that it can be parsed since it's not valid Javascript, so that's fine :-) The same happens with URL in a multiline comment without `*` at begin, so it's also covered by making check of comment headers optional since it can't be valid Javascript too (`^\\s*(?://|\\*)?\\s*\\S*[^:/?#]://[^?#]\\S*$`). I know there could be a false positive because `:` is used in conditionals and definitions, and `//` is a comment itself, but that's more of an issue of each particular case RegExp, not of `eslint` :-P

Maybe can we add a section on the `max-len` rule docs to add this kind of tricks and snippets for the `ignorePattern` property? :-)

**Comment by piranna** [comment]
I've updated my previous RegExp to `^\\s*(?://|\\*)?\\s*\\S*[^:/?#]://[^?#]\\S*\\s*$` so it can pass if the line has whitespaces at the end, so it will not fail if it's not trimmed.
